core/node/consensus/src/storage/connection.rs

```diff
@@ -239,17 +250,152 @@ impl<'a> Connection<'a> {
+        &mut self,
+        ctx: &ctx::Ctx,
+        number: attester::BatchNumber,
+    ) -> ctx::Result<Option<attester::SyncBatch>> {
+        let Some((min, max)) = self
+            .get_l2_block_range_of_l1_batch(ctx, number)
+            .await
+            .context("get_l2_block_range_of_l1_batch()")?
+        else {
+            return Ok(None);
+        };
+
+        let payloads = self.payloads(ctx, min..max).await.wrap("payloads()")?;
+        let payloads = payloads.into_iter().map(|p| p.encode()).collect();
+
+        // TODO: Fill out the proof when we have the stateless L1 batch validation story finished.
+        // It is supposed to be a Merkle proof that the rolling hash of the batch has been included
+        // in the L1 state tree. The state root hash of L1 won't be available in the DB, it requires
+        // an API client.
+        let batch = attester::SyncBatch {
+            number,
+            payloads,
+            proof: Vec::new(),
+        };
+
+        Ok(Some(batch))
+    }
+
+    /// Construct the [storage::BatchStoreState] which contains the earliest batch and the last available [attester::SyncBatch].
+    pub async fn batches_range(&mut self, ctx: &ctx::Ctx) -> ctx::Result<storage::BatchStoreState> {
+        let mut state = BatchStoreState {
+            first: attester::BatchNumber(0), // TODO: Is 0 okay for an empty database?
+            last: None,
+        };
+
+        if let Some(first) = self
+            .0
+            .blocks_dal()
+            .get_earliest_l1_batch_number()
+            .await
+            .context("get_earliest_l1_batch_number()")?
+        {
+            state.first.0 = first.0 as u64;
+        }
+
+        // TODO: In the future when we start filling in the `SyncBatch::proof` field,
+        // we can only run `get_batch` expecting `Some` result on numbers where the
+        // L1 state root hash is already available, so that we can produce some
+        // Merkle proof that the rolling hash of the L2 blocks in the batch has
+        // been included in the L1 state tree. At that point we probably can't
+        // call `get_last_batch_number` here, but something that indicates that
+        // the hashes/commitments on the L1 batch are ready and the thing has
+        // been included in L1; that potentially requires an API client as well.
+        if let Some(last) = self
+            .get_last_batch_number(ctx)
+            .await
+            .context("get_last_batch_number()")?
+        {
+            // For now it would be unexpected if we couldn't retrieve the payloads
+            // for the `last` batch number, as an L1 batch is only created if we
+            // have all the L2 miniblocks for it.
+            state.last = Some(
+                self.get_batch(ctx, last)
+                    .await
+                    .context("get_batch()")?
+                    .context("last batch not available")?,
+            );
+        }
+
+        Ok(state)
+    }
```

Comment:
nit: Is there any specific reason why it's done in imperative way? Could we use `unwrap_or` or similar to achieve the same behaviour?

Reply:
The reason was that it seemed to me that the `first` and `last` fields can independently be empty:
* Both can be empty if we haven't got any batch at all
* `first` can be determined but the inclusion on L1 means we might not be able to construct the proof yet, so `last` is empty
* eventually we'll have `last` for something, although it might not be what is currently the `get_last_batch_number`

I also needed to run 3 queries, all of which return `Option`, even though technically at the moment they should either all be `None` or all be `Some`. It's a bit messy, as the comments show it's not the final article, not even sure it should be in `Connection` eventually if it needs API clients. 

But there's no reason why it should be mutating `state`, I can put `unwrap_or`s at the end.